### PR TITLE
[Task] [68550] Build UI Search Screen

### PR DESCRIPTION
**Pull request by bachnd2k1**
### [Redmine link]
https://edu-redmine.sun-asterisk.vn/issues/68550

#### Screenshots (nếu có)

https://github.com/awesome-academy/ios_oe_nghiem_dinh_bach_project_1-Public/assets/75248050/e2c953a2-0bc1-416c-8ada-fb5bd01f1b99

### CODE REVIEW

File: MovieApp/MovieApp/Controller/Search/SearchViewController.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        tableView.dataSource = self
+        tableView.backgroundColor = .black
+        fetchDiscoverMovies()
+    }
+    
+    private func fetchDiscoverMovies() {
+        ApiCaller.shared.getDiscoverMovies { [weak self] results in
+            switch results {
+            case .success(let movies):
+                self?.movies = movies
```
Comment (ghost) [inline MovieApp/MovieApp/Controller/Search/SearchViewController.swift]: unwrap trước đi

File: MovieApp/MovieApp/Controller/Search/SearchViewController.swift [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        ApiCaller.shared.getVideoYoutube(with: name) { [weak self] result in
+            switch result {
+            case .success(let youtube):
+                DispatchQueue.main.async {
+                    let viewController = DetailViewController()
+                    viewController.config(youtube: youtube, movie: movie)
+                    self?.navigationController?.pushViewController(viewController, animated: true)
+                }
+            case .failure(let error):
+                self?.showAlertError(message: "error+ \(error)")
```
Comment (ghost) [inline MovieApp/MovieApp/Controller/Search/SearchViewController.swift]: xử lý self trước

File: MovieApp/MovieApp/Cell/Search/SearchCell.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+//
+//  Created by Bach Nghiem on 21/09/2023.
+//
+
+import UIKit
+protocol SearchCellDelegate: AnyObject {
+    func didTapItem(youtube: Youtube, movie: Movie)
+}
+
+class SearchCell: UITableViewCell {
```
Comment (hlinh88) [inline MovieApp/MovieApp/Cell/Search/SearchCell.swift]: final

File: MovieApp/MovieApp/Cell/Search/SearchCell.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    @IBOutlet private weak var previewImageView: UIImageView!
+    @IBOutlet private weak var playButton: UIButton!
+    @IBOutlet private weak var nameLabel: UILabel!
+    
+    private var movie: Movie?
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        // Initialization code
```
Comment (hlinh88) [inline MovieApp/MovieApp/Cell/Search/SearchCell.swift]: xoá

File: MovieApp/MovieApp/Controller/Search/SearchResultsViewController.swift [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+protocol SearchResultsViewControllerDelegate: AnyObject {
+    func didTapItem(youtube: Youtube, movie: Movie)
+    func showError(error: String)
+}
+
+final class SearchResultsViewController: UIViewController {
+    private let searchResultsCollectionView : UICollectionView = {
+        let layout = UICollectionViewFlowLayout()
+        layout.itemSize = CGSize(width: UIScreen.main.bounds.width / 3 - 10 , height: 200)
```
Comment (hlinh88) [inline MovieApp/MovieApp/Controller/Search/SearchResultsViewController.swift]: define magic numbers

File: MovieApp/MovieApp/Controller/Search/SearchViewController.swift [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+final class SearchViewController: UIViewController {
+    @IBOutlet private weak var tableView: UITableView!
+    private let searchController: UISearchController = {
+        let controller = UISearchController(searchResultsController: SearchResultsViewController())
+        controller.searchBar.placeholder = "Search for Movie or a Tv Show"
+        controller.searchBar.searchBarStyle = .minimal
+        controller.searchBar.searchTextField.backgroundColor = UIColor.lightGray
+        return controller
+    }()
```
Comment (hlinh88) [inline MovieApp/MovieApp/Controller/Search/SearchViewController.swift]:
```suggestion
    } ()
```

File: MovieApp/MovieApp/Controller/Search/SearchViewController.swift [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    private var movies = [Movie]()
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        title = "Search"
+        if let navigationController = navigationController {
+            navigationController.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.white]
+        }
+        navigationItem.searchController = searchController
+        navigationController?.navigationBar.tintColor = .white
+        searchController.searchResultsUpdater = self
```
Comment (hlinh88) [inline MovieApp/MovieApp/Controller/Search/SearchViewController.swift:33]: nên tách ra function riêng thay vì để hết viewDidLoad

**Review by ghost — APPROVED** [review]

**Pull request merged**